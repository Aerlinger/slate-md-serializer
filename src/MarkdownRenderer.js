import { Record } from 'immutable'

/**
 * String.
 */

const String = new Record({
  kind: 'string',
  text: ''
})

/**
 * A rule to (de)serialize text nodes. This is automatically added to the HTML
 * serializer so that users don't have to worry about text-level serialization.
 *
 * @type {Object}
 */

const TEXT_RULE = {
  serialize(obj, children) {
    if (obj.kind == 'string') {
      return children
    }
  }
}


/**
 * Markdown serializer.
 *
 * @type {Markdown}
 */

class Markdown {

  /**
   * Create a new serializer with `rules`.
   *
   * @param {Object} options
   *   @property {Array} rules
   * @return {Markdown} serializer
   */

  constructor(options = {}) {
    this.rules = [
      ...(options.rules || []),
      TEXT_RULE
    ];

    this.serializeNode = this.serializeNode.bind(this);
    this.serializeRange = this.serializeRange.bind(this);
    this.serializeString = this.serializeString.bind(this);
  }


  /**
   * Serialize a `state` object into an HTML string.
   *
   * @param {State} state
   * @return {String} markdown
   */

  serialize(state) {
    const { document } = state
    const elements = document.nodes.map(this.serializeNode)

    return elements.join('\n').trim();
  }

  /**
   * Serialize a `node`.
   *
   * @param {Node} node
   * @return {String}
   */

  serializeNode(node) {
    if (node.kind == 'text') {
      const ranges = node.getRanges()
      return ranges.map(this.serializeRange)
    }

    let children = node.nodes.map(this.serializeNode);
    children = children.flatten().length === 0 ? '' : children.flatten().join('')

    for (const rule of this.rules) {
      if (!rule.serialize) continue
      const ret = rule.serialize(node, children)
      if (ret) return ret
    }
  }

  /**
   * Serialize a `range`.
   *
   * @param {Range} range
   * @return {String}
   */

  serializeRange(range) {
    const string = new String({ text: range.text })
    const text = this.serializeString(string)

    return range.marks.reduce((children, mark) => {
      for (const rule of this.rules) {
        if (!rule.serialize) continue
        const ret = rule.serialize(mark, children)
        if (ret) return ret
      }
    }, text);
  }

  /**
   * Serialize a `string`.
   *
   * @param {String} string
   * @return {String}
   */

  serializeString(string) {
    for (const rule of this.rules) {
      if (!rule.serialize) continue
      const ret = rule.serialize(string, string.text)
      if (ret) return ret
    }
  }

}

/**
 * Export.
 */

export default Markdown
