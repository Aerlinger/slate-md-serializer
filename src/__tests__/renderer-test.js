import MarkdownRenderer from "../renderer";
const Markdown = new MarkdownRenderer();

// By parsing, rendering and reparsing we can test both sides of the serializer
// at the same time and ensure that parsing / rendering is compatible.
function getNodes(text) {
  const parsed = Markdown.deserialize(text);
  return parsed.document.nodes;

  const rendered = Markdown.serialize(parsed);
  const reparsed = Markdown.deserialize(rendered);
  return reparsed.document.nodes;
}

test("parses paragraph", () => {
  const output = Markdown.deserialize("This is just a sentance");
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading1", () => {
  const output = Markdown.deserialize("# Heading");
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading2", () => {
  const output = Markdown.deserialize("## Heading");
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading3", () => {
  const output = Markdown.deserialize("### Heading");
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading4", () => {
  const output = Markdown.deserialize("#### Heading");
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading5", () => {
  const output = Markdown.deserialize("##### Heading");
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading6", () => {
  const output = Markdown.deserialize("###### Heading");
  expect(output.document.nodes).toMatchSnapshot();
});

test("bold mark", () => {
  const text = `**this is bold**`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("italic mark", () => {
  const text = `*this is italic* _this is italic too_`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("deleted mark", () => {
  const text = `~~this is strikethrough~~`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("inserted mark", () => {
  const text = `++inserted text++`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("code mark", () => {
  const text = "`const foo = 123;`";
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses quote", () => {
  const text = `
> this is a quote
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses list items", () => {
  const text = `
- one
- two
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses list with trailing item", () => {
  const text = `
- one
- two
-
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses indented list items", () => {
  const text = `
 - one
 - two
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses list items with marks", () => {
  const text = `
 - one **bold**
 - *italic* two
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses ordered list items", () => {
  const text = `
1. one
1. two
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses ordered list with trailing item", () => {
  const text = `
1. one
1. two
1.
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses ordered list items with marks", () => {
  const text = `
1. one **bold**
1. *italic* two
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses ordered list items with different numbers", () => {
  const text = `
1. one
2. two
3. three
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses mixed list items", () => {
  const text = `
1. list

- another

1. different
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses tables", () => {
  const text = `
| Tables   |      Are      |  Cool |
|----------|:-------------:|------:|
| col 1 is |  left-aligned | $1600 |
| col 2 is |    centered   |   $12 |
| col 3 is | right-aligned |    $1 |
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses todo list items", () => {
  const text = `
[ ] todo
[x] done
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses nested todo list items", () => {
  const text = `
[ ] todo
   [ ] nested
   [ ] deep
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses double nested todo list items", () => {
  const text = `
[x] checked
   [ ] empty
   [x] checked

[ ] three
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses todo list items with marks", () => {
  const text = `
 [x] ~~done~~
 [x] more **done**
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses ``` code fences", () => {
  const text = `
\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses ~~~ code fences", () => {
  const text = `
~~~
const hello = 'world';
function() {
  return hello;
}
~~~
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses indented code blocks", () => {
  const text = `
    const hello = 'world';
    function() {
      return hello;
    }
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses link", () => {
  const text = `[google](http://google.com)`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses link within mark", () => {
  const text = `**[google](http://google.com)**`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses link with encoded characters", () => {
  const text = `[kibana](https://example.com/app/kibana#/discover?_g=%28refreshInterval:%28%27$$hashKey%27:%27object:1596%27,display:%2710%20seconds%27,pause:!f,section:1,value:10000%29,time:%28from:now-15m,mode:quick,to:now%29%29&_a=%28columns:!%28metadata.step,message,metadata.attempt_f,metadata.tries_f,metadata.error_class,metadata.url%29,index:%27logs-%27,interval:auto,query:%28query_string:%28analyze_wildcard:!t,query:%27metadata.at:%20Stepper*%27%29%29,sort:!%28time,desc%29%29)`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses link with percent symbol", () => {
  const text = `[kibana](https://example.com/app/kibana#/visualize/edit/Requests-%)`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses interesting nesting", () => {
  const text = `
* List item that contains a blockquote with inline mark

  > Blockquote with code \`mapStateToProps()\`
`;
  expect(getNodes(text)).toMatchSnapshot();
});

test("parses empty string", () => {
  expect(getNodes("")).toMatchSnapshot();
});

test("parses whitespace string", () => {
  expect(getNodes("   ")).toMatchSnapshot();
});
