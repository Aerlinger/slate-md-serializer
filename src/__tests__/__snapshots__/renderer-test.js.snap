// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses \`\`\` code fences 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "const hello = 'world';
function() {
  return hello;
}",
          },
        ],
      },
    ],
    "type": "code",
  },
]
`;

exports[`parses ~~~ code fences 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "const hello = 'world';
function() {
  return hello;
}",
          },
        ],
      },
    ],
    "type": "code",
  },
]
`;

exports[`parses double nested todo list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {
          "checked": true,
        },
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "checked",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {
          "checked": false,
        },
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "three",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "todo-list",
  },
]
`;

exports[`parses empty string 1`] = `Array []`;

exports[`parses heading1 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "Heading",
          },
        ],
      },
    ],
    "type": "heading1",
  },
]
`;

exports[`parses heading2 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "Heading",
          },
        ],
      },
    ],
    "type": "heading2",
  },
]
`;

exports[`parses heading3 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "Heading",
          },
        ],
      },
    ],
    "type": "heading3",
  },
]
`;

exports[`parses heading4 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "Heading",
          },
        ],
      },
    ],
    "type": "heading4",
  },
]
`;

exports[`parses heading5 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "Heading",
          },
        ],
      },
    ],
    "type": "heading5",
  },
]
`;

exports[`parses heading6 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "Heading",
          },
        ],
      },
    ],
    "type": "heading6",
  },
]
`;

exports[`parses indented code blocks 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "const hello = 'world';
function() {
  return hello;
}",
          },
        ],
      },
    ],
    "type": "code",
  },
]
`;

exports[`parses indented list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "one",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "two",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
]
`;

exports[`parses interesting nesting 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "List item that contains a blockquote with inline mark",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
]
`;

exports[`parses link 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "",
          },
        ],
      },
      Object {
        "data": Object {
          "href": "http://google.com",
        },
        "isVoid": false,
        "kind": "inline",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "google",
              },
            ],
          },
        ],
        "type": "link",
      },
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "",
          },
        ],
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parses link with encoded characters 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "",
          },
        ],
      },
      Object {
        "data": Object {
          "href": "https://example.com/app/kibana#/discover?_g=(refreshInterval:('$$hashKey':'object:1596',display:'10 seconds',pause:!f,section:1,value:10000),time:(from:now-15m,mode:quick,to:now))&_a=(columns:!(metadata.step,message,metadata.attempt_f,metadata.tries_f,metadata.error_class,metadata.url),index:'logs-',interval:auto,query:(query_string:(analyze_wildcard:!t,query:'metadata.at: Stepper*')),sort:!(time,desc))",
        },
        "isVoid": false,
        "kind": "inline",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "kibana",
              },
            ],
          },
        ],
        "type": "link",
      },
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "",
          },
        ],
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parses link within mark 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "",
          },
        ],
      },
      Object {
        "data": Object {
          "href": "http://google.com",
        },
        "isVoid": false,
        "kind": "inline",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "google",
              },
            ],
          },
        ],
        "type": "link",
      },
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "",
          },
        ],
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parses list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "one",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "two",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
]
`;

exports[`parses list items with marks 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "one ",
              },
              Object {
                "kind": "leaf",
                "marks": Array [
                  Object {
                    "data": Object {},
                    "kind": "mark",
                    "type": "bold",
                  },
                ],
                "text": "bold",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [
                  Object {
                    "data": Object {},
                    "kind": "mark",
                    "type": "italic",
                  },
                ],
                "text": "italic",
              },
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": " two",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
]
`;

exports[`parses list with trailing item 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "one",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "two
-",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
]
`;

exports[`parses mixed list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "list",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "ordered-list",
  },
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "another",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "different",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "ordered-list",
  },
]
`;

exports[`parses nested todo list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {
          "checked": false,
        },
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "todo",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "todo-list",
  },
]
`;

exports[`parses ordered list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "one",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "two",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "ordered-list",
  },
]
`;

exports[`parses ordered list items with different numbers 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "one",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "two",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "three",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "ordered-list",
  },
]
`;

exports[`parses ordered list items with marks 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "one ",
              },
              Object {
                "kind": "leaf",
                "marks": Array [
                  Object {
                    "data": Object {},
                    "kind": "mark",
                    "type": "bold",
                  },
                ],
                "text": "bold",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [
                  Object {
                    "data": Object {},
                    "kind": "mark",
                    "type": "italic",
                  },
                ],
                "text": "italic",
              },
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": " two",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "ordered-list",
  },
]
`;

exports[`parses paragraph 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "This is just a sentance",
          },
        ],
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parses quote 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": "this is a quote",
          },
        ],
      },
    ],
    "type": "block-quote",
  },
]
`;

exports[`parses quote with marks 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "kind": "text",
        "leaves": Array [
          Object {
            "kind": "leaf",
            "marks": Array [
              Object {
                "data": Object {},
                "kind": "mark",
                "type": "bold",
              },
            ],
            "text": "bold",
          },
          Object {
            "kind": "leaf",
            "marks": Array [],
            "text": " in a quote",
          },
        ],
      },
    ],
    "type": "block-quote",
  },
]
`;

exports[`parses tables 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "data": Object {
              "align": null,
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "Tables",
                  },
                ],
              },
            ],
            "type": "table-head",
          },
          Object {
            "data": Object {
              "align": "center",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "Are",
                  },
                ],
              },
            ],
            "type": "table-head",
          },
          Object {
            "data": Object {
              "align": "right",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "Cool",
                  },
                ],
              },
            ],
            "type": "table-head",
          },
        ],
        "type": "table-row",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "data": Object {
              "align": null,
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "col 1 is",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
          Object {
            "data": Object {
              "align": "center",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "left-aligned",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
          Object {
            "data": Object {
              "align": "right",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "$1600",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
        ],
        "type": "table-row",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "data": Object {
              "align": null,
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "col 2 is",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
          Object {
            "data": Object {
              "align": "center",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "centered",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
          Object {
            "data": Object {
              "align": "right",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "$12",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
        ],
        "type": "table-row",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "data": Object {
              "align": null,
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "col 3 is",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
          Object {
            "data": Object {
              "align": "center",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "right-aligned",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
          Object {
            "data": Object {
              "align": "right",
            },
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "kind": "text",
                "leaves": Array [
                  Object {
                    "kind": "leaf",
                    "marks": Array [],
                    "text": "$1",
                  },
                ],
              },
            ],
            "type": "table-cell",
          },
        ],
        "type": "table-row",
      },
    ],
    "type": "table",
  },
]
`;

exports[`parses todo list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {
          "checked": false,
        },
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "todo",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {
          "checked": true,
        },
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "done",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "todo-list",
  },
]
`;

exports[`parses todo list items with marks 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "kind": "block",
    "nodes": Array [
      Object {
        "data": Object {
          "checked": true,
        },
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [
                  Object {
                    "data": Object {},
                    "kind": "mark",
                    "type": "deleted",
                  },
                ],
                "text": "done",
              },
            ],
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {
          "checked": true,
        },
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "kind": "text",
            "leaves": Array [
              Object {
                "kind": "leaf",
                "marks": Array [],
                "text": "more ",
              },
              Object {
                "kind": "leaf",
                "marks": Array [
                  Object {
                    "data": Object {},
                    "kind": "mark",
                    "type": "bold",
                  },
                ],
                "text": "done",
              },
            ],
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "todo-list",
  },
]
`;

exports[`parses whitespace string 1`] = `Array []`;
